// @microsoft/teamsfx-core/build/plugins/resource/frontend/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const constants_1 = require("../constants");
class Logger {
    static setLogger(_logger) {
        this.logger = _logger;
    }
    static debug(message) {
        var _a;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.debug(`[${constants_1.FrontendPluginInfo.DisplayName}] ${message}`);
    }
    static info(message) {
        var _a;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(`[${constants_1.FrontendPluginInfo.DisplayName}] ${message}`);
    }
    static warning(message) {
        var _a;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.warning(`[${constants_1.FrontendPluginInfo.DisplayName}] ${message}`);
    }
    static error(message) {
        var _a;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.error(`[${constants_1.FrontendPluginInfo.DisplayName}] ${message}`);
    }
}
exports.Logger = Logger;
