// @microsoft/teamsfx-core/build/plugins/resource/frontend/blazor/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceType = exports.AppSettingsKey = exports.BlazorConfigKey = void 0;
var BlazorConfigKey;
(function (BlazorConfigKey) {
    /* Config from solution */
    BlazorConfigKey["resourceGroupName"] = "resourceGroupName";
    BlazorConfigKey["subscriptionId"] = "subscriptionId";
    BlazorConfigKey["resourceNameSuffix"] = "resourceNameSuffix";
    BlazorConfigKey["location"] = "location";
    BlazorConfigKey["credential"] = "credential";
    BlazorConfigKey["teamsAppName"] = "teamsAppName";
    /* Config exported by Blazor plugin */
    BlazorConfigKey["webAppName"] = "webAppName";
    BlazorConfigKey["appServicePlanName"] = "appServicePlanName";
    BlazorConfigKey["webAppEndpoint"] = "webAppEndpoint";
    BlazorConfigKey["webAppDomain"] = "webAppDomain";
    BlazorConfigKey["projectFilePath"] = "projectFilePath";
    BlazorConfigKey["webAppResourceId"] = "webAppResourceId";
    /* Intermediate */
    BlazorConfigKey["site"] = "site";
})(BlazorConfigKey = exports.BlazorConfigKey || (exports.BlazorConfigKey = {}));
var AppSettingsKey;
(function (AppSettingsKey) {
    AppSettingsKey["clientSecret"] = "CLIENT_SECRET";
    AppSettingsKey["clientId"] = "CLIENT_ID";
    AppSettingsKey["oauthHost"] = "OAUTH_AUTHORITY";
    AppSettingsKey["tabAppEndpoint"] = "TAB_APP_ENDPOINT";
    AppSettingsKey["aadMetadataAddress"] = "AAD_METADATA_ADDRESS";
    AppSettingsKey["botId"] = "BOT_ID";
    AppSettingsKey["botPassword"] = "BOT_PASSWORD";
    AppSettingsKey["identifierUri"] = "IDENTIFIER_URI";
})(AppSettingsKey = exports.AppSettingsKey || (exports.AppSettingsKey = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["webApp"] = "Azure Web App";
    ResourceType["appServicePlan"] = "Azure App Service Plan";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
