// @microsoft/teamsfx-core/build/plugins/resource/bot/resources/strings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientNames = exports.Commands = exports.ConfigNames = exports.TelemetryStrings = exports.PluginBot = exports.PluginFunction = exports.PluginIdentity = exports.PluginSql = exports.PluginSolution = exports.PluginLocalDebug = exports.PluginAAD = exports.CommonStrings = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
class CommonStrings {
}
exports.CommonStrings = CommonStrings;
CommonStrings.BOT_WORKING_DIR_NAME = "bot";
CommonStrings.DEFAULT_FILE_ENCODING = "utf-8";
// for the use of migrating v1 project
CommonStrings.NODE_PACKAGE_FOLDER_NAME = "node_modules";
CommonStrings.APP_SERVICE_PLAN = "App Service plan";
CommonStrings.AZURE_WEB_APP = "Azure Web App";
CommonStrings.BOT_CHANNEL_REGISTRATION = "Azure Bot Service channel registration";
CommonStrings.MS_TEAMS_CHANNEL = "Teams channel";
CommonStrings.AAD_APP = "AAD app";
CommonStrings.AAD_CLIENT_SECRET = "AAD client secret";
CommonStrings.APPSTUDIO_BOT_REGISTRATION = "App Studio bot registration";
CommonStrings.APPSTUDIO_MSG_ENDPOINT = "App Studio message endpoint";
CommonStrings.CONFIG_ITEM = "config item";
CommonStrings.SHORT_APP_NAME = "short app name";
CommonStrings.AUTH_REDIRECT_URI_SUFFIX = "/public";
CommonStrings.AUTH_LOGIN_URI_SUFFIX = "/auth-start.html";
CommonStrings.HTTPS_PREFIX = "https://";
CommonStrings.MESSAGE_ENDPOINT_SUFFIX = "/api/messages";
class PluginAAD {
}
exports.PluginAAD = PluginAAD;
PluginAAD.PLUGIN_NAME = "fx-resource-aad-app-for-teams";
PluginAAD.CLIENT_ID = "clientId";
PluginAAD.CLIENT_SECRET = "clientSecret";
PluginAAD.APPLICATION_ID_URIS = "applicationIdUris";
class PluginLocalDebug {
}
exports.PluginLocalDebug = PluginLocalDebug;
PluginLocalDebug.PLUGIN_NAME = "fx-resource-local-debug";
PluginLocalDebug.LOCAL_BOT_ENDPOINT = "localBotEndpoint";
PluginLocalDebug.LOCAL_DEBUG_SUFFIX = "-local-debug";
class PluginSolution {
}
exports.PluginSolution = PluginSolution;
PluginSolution.PLUGIN_NAME = "solution";
PluginSolution.SUBSCRIPTION_ID = "subscriptionId";
PluginSolution.RESOURCE_GROUP_NAME = "resourceGroupName";
PluginSolution.LOCATION = "location";
PluginSolution.M365_TENANT_ID = "teamsAppTenantId";
PluginSolution.RESOURCE_NAME_SUFFIX = "resourceNameSuffix";
PluginSolution.REMOTE_TEAMS_APPID = "remoteTeamsAppId";
class PluginSql {
}
exports.PluginSql = PluginSql;
PluginSql.PLUGIN_NAME = "fx-resource-azure-sql";
PluginSql.SQL_ENDPOINT = "sqlEndpoint";
PluginSql.SQL_DATABASE_NAME = "databaseName";
PluginSql.SQL_USERNAME = "sqlUsername";
PluginSql.SQL_PASSWORD = "sqlPassword";
class PluginIdentity {
}
exports.PluginIdentity = PluginIdentity;
PluginIdentity.PLUGIN_NAME = "fx-resource-identity";
PluginIdentity.IDENTITY_ClIENT_ID = "identityClientId";
PluginIdentity.IDENTITY_RESOURCE_ID = "identityResourceId";
class PluginFunction {
}
exports.PluginFunction = PluginFunction;
PluginFunction.PLUGIN_NAME = "fx-resource-function";
PluginFunction.ENDPOINT = "functionEndpoint";
class PluginBot {
}
exports.PluginBot = PluginBot;
PluginBot.PLUGIN_NAME = "fx-resource-bot";
PluginBot.BOT_ID = "botId";
PluginBot.BOT_PASSWORD = "botPassword";
PluginBot.OBJECT_ID = "objectId";
PluginBot.LOCAL_BOT_ID = "localBotId";
PluginBot.LOCAL_BOT_PASSWORD = "localBotPassword";
PluginBot.LOCAL_OBJECT_ID = "localObjectId";
PluginBot.PROGRAMMING_LANGUAGE = "programmingLanguage";
PluginBot.SCAFFOLDED = "scaffolded";
PluginBot.APP_SERVICE_PLAN = "appServicePlan";
PluginBot.SITE_NAME = "siteName";
PluginBot.SKU_NAME = "skuName";
PluginBot.SITE_ENDPOINT = "siteEndpoint";
PluginBot.VALID_DOMAIN = "validDomain";
PluginBot.PROVISIONED = "provisioned";
PluginBot.WEB_APPLICATION_INFO_ID = "webApplicationInfo.id";
PluginBot.WEB_APPLICATION_INFO_RESOURCE = "webApplicationInfo.resource";
PluginBot.BOTS_SECTION = "bots";
PluginBot.BOT_CHANNEL_REGISTRATION = "botChannelReg";
PluginBot.BOT_WEB_APP_RESOURCE_ID = "botWebAppResourceId";
PluginBot.UNPACK_FLAG = "unPackFlag";
PluginBot.MESSAGE_EXTENSION_SECTION = "composeExtensions";
PluginBot.LOCAL_REDIRECT_URI = "local_redirectUri";
class TelemetryStrings {
}
exports.TelemetryStrings = TelemetryStrings;
TelemetryStrings.COMPONENT_NAME = "fx-resource-bot";
class ConfigNames {
}
exports.ConfigNames = ConfigNames;
ConfigNames.PROGRAMMING_LANGUAGE = "programming language";
ConfigNames.GRAPH_TOKEN = "graph token";
ConfigNames.SUBSCRIPTION_ID = "subscription id";
ConfigNames.SERVICE_CLIENT_CREDENTIALS = "service client credentials";
ConfigNames.LOCATION = "location";
ConfigNames.SKU_NAME = "sku name";
ConfigNames.BOT_SERVICE_RESOURCE_ID = "bot service resource id";
ConfigNames.RESOURCE_GROUP = "resource group";
ConfigNames.LOCAL_ENDPOINT = "local endpoint";
ConfigNames.AUTH_CLIENT_ID = "auth client id";
ConfigNames.AUTH_CLIENT_SECRET = "auth client secret";
ConfigNames.AUTH_TENANT = "auth tenant";
ConfigNames.AUTH_APPLICATION_ID_URIS = "auth application id uris";
ConfigNames.SITE_ENDPOINT = "site endpoint";
ConfigNames.BOT_ID = "bot id";
ConfigNames.BOT_PASSWORD = "bot password";
ConfigNames.LOCAL_BOT_ID = "local bot id";
ConfigNames.LOCAL_BOT_PASSWORD = "local bot password";
ConfigNames.APPSTUDIO_TOKEN = "app studio token";
ConfigNames.AZURE_WEB_APP_AUTH_CONFIGS = "azure web app's auth configs";
ConfigNames.MESSAGE_ENDPOINT = "message endpoint";
class Commands {
}
exports.Commands = Commands;
Commands.NPM_INSTALL = "npm install";
Commands.NPM_BUILD = "npm run build";
class ClientNames {
}
exports.ClientNames = ClientNames;
ClientNames.WEB_SITE_MGMT_CLIENT = "webSiteMgmtClient";
ClientNames.BOT_SERVICE_CLIENT = "botServiceClient";
