// @microsoft/teamsfx-core/build/plugins/solution/spfx-solution/scaffold.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.addResource = exports.publishApplication = exports.generateResourceTemplate = exports.scaffold = exports.getQuestionsForScaffold = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const typedi_1 = tslib_1.__importDefault(require("typedi"));
const constants_1 = require("../fx-solution/v3/constants");
const error_1 = require("../utils/error");
const questions_1 = require("../utils/questions");
async function getQuestionsForScaffold(ctx, inputs) {
    const node = new teamsfx_api_1.QTreeNode({ type: "group" });
    const solutionSettings = ctx.projectSetting.solutionSettings;
    if (solutionSettings.modules) {
        const moduleNode = questions_1.createSelectModuleQuestionNode(solutionSettings.modules);
        node.addChild(moduleNode);
    }
    const templateNode = new teamsfx_api_1.QTreeNode(questions_1.selectScaffoldTemplateQuestion);
    const staticOptions = [];
    const spfxPlugin = typedi_1.default.get(constants_1.BuiltInScaffoldPluginNames.spfx);
    const scaffoldPlugins = [spfxPlugin];
    for (const plugin of scaffoldPlugins) {
        const getTemplatesRes = await plugin.getTemplates(ctx, inputs);
        if (getTemplatesRes.isErr()) {
            return teamsfx_api_1.err(getTemplatesRes.error);
        }
        for (const template of getTemplatesRes.value) {
            staticOptions.push({
                id: `${plugin.name}/${template.name}`,
                label: `${template.name}(${template.language})`,
                detail: template.description,
                data: {
                    pluginName: plugin.name,
                    templateName: template.name,
                },
            });
        }
        if (plugin.getQuestionsForScaffold) {
            const pluginQuestionsRes = await plugin.getQuestionsForScaffold(ctx, inputs);
            if (pluginQuestionsRes.isOk()) {
                const pluginNode = pluginQuestionsRes.value;
                if (pluginNode) {
                    pluginNode.condition = {
                        validFunc: async (input, inputs) => {
                            if (input.data) {
                                if (input.data.pluginName === plugin.name)
                                    return undefined;
                            }
                            return "";
                        },
                    };
                    templateNode.addChild(pluginNode);
                }
            }
        }
    }
    questions_1.selectScaffoldTemplateQuestion.staticOptions = staticOptions;
    node.addChild(templateNode);
    return teamsfx_api_1.ok(node);
}
exports.getQuestionsForScaffold = getQuestionsForScaffold;
async function scaffold(ctx, inputs) {
    if (!inputs.template) {
        return teamsfx_api_1.err(new error_1.InvalidInputError(inputs));
    }
    const template = inputs.template;
    if (!template.data) {
        return teamsfx_api_1.err(new error_1.InvalidInputError(inputs, "template.data is undefined"));
    }
    const data = template.data;
    const pluginName = data.pluginName;
    const templateName = data.templateName;
    const plugin = typedi_1.default.get(pluginName);
    const pluginInputs = Object.assign(Object.assign({}, inputs), { template: templateName });
    const res = await plugin.scaffold(ctx, pluginInputs);
    if (res.isErr()) {
        return teamsfx_api_1.err(res.error);
    }
    const manifest = [];
    if (res.value) {
        manifest.push(res.value);
    }
    inputs.manifest = manifest;
    pluginInputs.manifest = manifest;
    //TODO
    // //call appstudio.scaffold() API
    // const appstudioPlugin = Container.get<v3.ScaffoldPlugin>(BuiltInResourcePluginNames.AppStudio);
    // await appstudioPlugin.scaffold(ctx, pluginInputs);
    return teamsfx_api_1.ok(teamsfx_api_1.Void);
}
exports.scaffold = scaffold;
async function generateResourceTemplate(ctx, inputs) {
    return teamsfx_api_1.ok({});
}
exports.generateResourceTemplate = generateResourceTemplate;
async function publishApplication(ctx, inputs, envInfo, tokenProvider) {
    return teamsfx_api_1.ok(teamsfx_api_1.Void);
}
exports.publishApplication = publishApplication;
async function addResource(ctx, inputs) {
    return teamsfx_api_1.ok(teamsfx_api_1.Void);
}
exports.addResource = addResource;
