// @microsoft/teamsfx-core/build/plugins/solution/spfx-solution/solution.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsSPFxSolution = void 0;
const tslib_1 = require("tslib");
const constants_1 = require("../../../common/constants");
const scaffold_1 = require("./scaffold");
const init_1 = require("./init");
const typedi_1 = require("typedi");
const constants_2 = require("../fx-solution/v3/constants");
const addModule_1 = require("./addModule");
let TeamsSPFxSolution = class TeamsSPFxSolution {
    constructor() {
        this.name = constants_2.BuiltInSolutionNames.spfx;
        this.displayName = constants_1.PluginDisplayName.SpfxSolution;
        this.init = init_1.init;
        this.getQuestionsForInit = init_1.getQuestionsForInit;
        this.scaffold = scaffold_1.scaffold;
        this.getQuestionsForScaffold = scaffold_1.getQuestionsForScaffold;
        this.generateResourceTemplate = scaffold_1.generateResourceTemplate;
        this.publishApplication = scaffold_1.publishApplication;
        this.addResource = scaffold_1.addResource;
        this.addModule = addModule_1.addModule;
    }
};
TeamsSPFxSolution = tslib_1.__decorate([
    typedi_1.Service(constants_2.BuiltInSolutionNames.spfx)
], TeamsSPFxSolution);
exports.TeamsSPFxSolution = TeamsSPFxSolution;
