// @microsoft/teamsfx-core/build/plugins/resource/spfx/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployProgressMessage = exports.PreDeployProgressMessage = exports.ProgressTitleMessage = exports.TelemetryEvent = exports.TelemetryValue = exports.TelemetryKey = exports.PlaceHolders = exports.Constants = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
class Constants {
}
exports.Constants = Constants;
Constants.FRAMEWORK_NONE = "none";
Constants.FRAMEWORK_REACT = "react";
Constants.MAX_ALIAS_LENGTH = 40;
Constants.MAX_BUNDLE_NAME_LENGTH = 64;
Constants.CALLED_ID = "teamsdev";
Constants.APP_CATALOG_REFRESH_TIME = 10000;
Constants.APP_CATALOG_MAX_TIMES = 12;
Constants.PLUGIN_NAME = "SPFx";
Constants.PLUGIN_DEV_NAME = "fx-resource-spfx";
Constants.BUILD_SHAREPOINT_PACKAGE = "Build SharePoint Package";
Constants.READ_MORE = "Read more";
Constants.CANCEL = "Cancel";
Constants.DEPLOY_GUIDE = "https://docs.microsoft.com/en-us/microsoftteams/platform/get-started/first-app-spfx?tabs=vscode#deploy-your-app-to-sharepoint";
Constants.CREATE_APP_CATALOG_GUIDE = "https://docs.microsoft.com/en-us/sharepoint/use-app-catalog#create-the-app-catalog";
class PlaceHolders {
}
exports.PlaceHolders = PlaceHolders;
PlaceHolders.componentName = "<%= componentName %>";
PlaceHolders.componentNameCamelCase = "<%= componentNameCamelCase %>";
PlaceHolders.componentClassName = "<%= componentClassName %>";
PlaceHolders.componentStrings = "<%= componentStrings %>";
PlaceHolders.libraryName = "<%= libraryName %>";
PlaceHolders.componentId = "<%= componentId %>";
PlaceHolders.componentAlias = "<%= componentAlias %>";
PlaceHolders.componentDescription = "<%= componentDescription %>";
PlaceHolders.componentNameUnescaped = "<%= componentNameUnescaped %>";
PlaceHolders.componentClassNameKebabCase = "<%= componentClassNameKebabCase %>";
class TelemetryKey {
}
exports.TelemetryKey = TelemetryKey;
TelemetryKey.Component = "component";
TelemetryKey.Success = "success";
TelemetryKey.ErrorType = "error-type";
TelemetryKey.ErrorMessage = "error-message";
class TelemetryValue {
}
exports.TelemetryValue = TelemetryValue;
TelemetryValue.Success = "yes";
TelemetryValue.Fail = "no";
TelemetryValue.UserError = "user";
TelemetryValue.SystemError = "system";
class TelemetryEvent {
}
exports.TelemetryEvent = TelemetryEvent;
TelemetryEvent.StartSuffix = "-start";
TelemetryEvent.Scaffold = "scaffold";
TelemetryEvent.PreDeploy = "pre-deploy";
TelemetryEvent.Deploy = "deploy";
class ProgressTitleMessage {
}
exports.ProgressTitleMessage = ProgressTitleMessage;
ProgressTitleMessage.PreDeployProgressTitle = "Building SharePoint package";
ProgressTitleMessage.DeployProgressTitle = `Upload and deploy SharePoint Package`;
class PreDeployProgressMessage {
}
exports.PreDeployProgressMessage = PreDeployProgressMessage;
PreDeployProgressMessage.NpmInstall = "Run: npm install. This may take more than 5 minutes to finish";
PreDeployProgressMessage.GulpBundle = "Run: gulp bundle --ship";
PreDeployProgressMessage.GulpPackage = "Run: gulp package-solution --ship";
class DeployProgressMessage {
}
exports.DeployProgressMessage = DeployProgressMessage;
DeployProgressMessage.GetSPAppCatalog = "Get SharePoint app catalog";
DeployProgressMessage.UploadAndDeploy = "Upload and deploy SPFx package on your tenant website";
