// @microsoft/teamsfx-core/build/plugins/resource/spfx/utils/progress-helper.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressHelper = void 0;
const constants_1 = require("./constants");
class ProgressHelper {
    static async startPreDeployProgressHandler(ctx) {
        var _a, _b;
        this.preDeployProgress = (_a = ctx.ui) === null || _a === void 0 ? void 0 : _a.createProgressBar(constants_1.ProgressTitleMessage.PreDeployProgressTitle, Object.entries(constants_1.PreDeployProgressMessage).length);
        await ((_b = this.preDeployProgress) === null || _b === void 0 ? void 0 : _b.start(""));
        return this.preDeployProgress;
    }
    static async endAllHandlers(success) {
        await this.endPreDeployProgress(success);
    }
    static async endPreDeployProgress(success) {
        var _a;
        await ((_a = this.preDeployProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.preDeployProgress = undefined;
    }
    static async startDeployProgressHandler(ctx) {
        var _a, _b;
        this.deployProgress = (_a = ctx.ui) === null || _a === void 0 ? void 0 : _a.createProgressBar(constants_1.ProgressTitleMessage.DeployProgressTitle, 1);
        await ((_b = this.deployProgress) === null || _b === void 0 ? void 0 : _b.start(""));
        return this.deployProgress;
    }
    static async endDeployProgress(success) {
        var _a;
        await ((_a = this.deployProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.deployProgress = undefined;
    }
}
exports.ProgressHelper = ProgressHelper;
