// @microsoft/teamsfx-core/build/plugins/resource/sql/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureSqlBicepFile = exports.AzureSqlBicep = exports.HelpLinks = exports.Telemetry = exports.Constants = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
class Constants {
}
exports.Constants = Constants;
Constants.pluginName = "SQL Plugin";
Constants.pluginNameShort = "SQL";
Constants.pluginFullName = "fx-resource-azure-sql";
Constants.firewall = {
    azureIp: "0.0.0.0",
    azureRule: "AllowAzure",
    localRule: "AllowLocal",
};
Constants.ipBeginToken = "0";
Constants.ipEndToken = "255";
Constants.maxRetryTimes = 3;
Constants.jwtToken = {
    ver1: "1.0",
    ver2: "2.0",
    userType: "0",
};
Constants.sqlAdministratorType = "ActiveDirectory";
Constants.echoIpAddress = "https://api.ipify.org";
Constants.azureSqlScope = "https://database.windows.net//.default";
Constants.resourceGroupName = "resourceGroupName";
Constants.sqlEndpoint = "sqlEndpoint";
Constants.sqlResourceId = "sqlResourceId";
Constants.databaseName = "databaseName";
Constants.skipAddingSqlUser = "skipAddingSqlUser";
Constants.admin = "admin";
Constants.adminPassword = "adminPassword";
Constants.solution = "solution";
Constants.solutionPluginFullName = "fx-solution-azure";
Constants.solutionConfigKey = {
    subscriptionId: "subscriptionId",
    resourceGroupName: "resourceGroupName",
    resourceNameSuffix: "resourceNameSuffix",
    location: "location",
    tenantId: "tenantId",
    remoteTeamsAppId: "remoteTeamsAppId",
};
Constants.identityPlugin = "fx-resource-identity";
Constants.identityName = "identityName";
Constants.userQuestion = {
    adminName: "Admin name of SQL",
    adminPassword: "Admin password of SQL",
    confirmPassword: "Confirm admin password of SQL",
    skipAddingUser: "Skip adding database user",
};
Constants.questionKey = {
    adminName: "sql-admin-name",
    adminPassword: "sql-password",
    confirmPassword: "sql-confirm-password",
    skipAddingUser: "sql-skip-adding-user",
};
Constants.resourceProvider = "Microsoft.Sql";
class Telemetry {
}
exports.Telemetry = Telemetry;
Telemetry.componentName = "fx-resource-azure-sql";
Telemetry.startSuffix = "-start";
Telemetry.valueYes = "yes";
Telemetry.valueNo = "no";
Telemetry.userError = "user";
Telemetry.systemError = "system";
Telemetry.stage = {
    preProvision: "pre-provision",
    provision: "provision",
    postProvision: "post-provision",
    getQuestion: "get-question",
    generateArmTemplates: "generate-arm-templates",
};
Telemetry.properties = {
    component: "component",
    success: "success",
    errorCode: "error-code",
    errorType: "error-type",
    errorMessage: "error-message",
    appid: "appid",
    skipAddingUser: "skip-adding-user",
};
class HelpLinks {
}
exports.HelpLinks = HelpLinks;
HelpLinks.default = "https://aka.ms/teamsfx-sql-help";
class AzureSqlBicep {
}
exports.AzureSqlBicep = AzureSqlBicep;
AzureSqlBicep.sqlResourceId = "provisionOutputs.azureSqlOutput.value.sqlResourceId";
AzureSqlBicep.sqlEndpoint = "provisionOutputs.azureSqlOutput.value.sqlEndpoint";
AzureSqlBicep.databaseName = "provisionOutputs.azureSqlOutput.value.databaseName";
class AzureSqlBicepFile {
}
exports.AzureSqlBicepFile = AzureSqlBicepFile;
AzureSqlBicepFile.moduleTemplateFileName = "provision.template.bicep";
AzureSqlBicepFile.ProvisionModuleTemplateFileName = "sqlProvision.template.bicep";
