// @microsoft/teamsfx-core/build/plugins/resource/function/utils/depsChecker/funcPluginLogger.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.funcPluginLogger = void 0;
const tslib_1 = require("tslib");
const os = tslib_1.__importStar(require("os"));
const logger_1 = require("../logger");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
class FuncPluginLogger {
    constructor() {
        this.detailLogLines = [];
    }
    debug(message) {
        this.appendLine(teamsfx_api_1.LogLevel.Debug, message);
        return Promise.resolve(true);
    }
    info(message) {
        this.appendLine(teamsfx_api_1.LogLevel.Info, message);
        logger_1.Logger.info(message);
        return Promise.resolve(true);
    }
    warning(message) {
        this.appendLine(teamsfx_api_1.LogLevel.Warning, message);
        logger_1.Logger.warning(message);
        return Promise.resolve(true);
    }
    async error(message) {
        this.appendLine(teamsfx_api_1.LogLevel.Error, message);
        logger_1.Logger.error(message);
        return Promise.resolve(true);
    }
    async printDetailLog() {
        logger_1.Logger.error(this.detailLogLines.join(os.EOL));
    }
    cleanup() {
        this.detailLogLines = [];
    }
    appendLine(level, message) {
        const line = `${teamsfx_api_1.LogLevel[level]} ${new Date().toISOString()}: ${message}`;
        this.detailLogLines.push(line);
    }
}
exports.funcPluginLogger = new FuncPluginLogger();
