// @microsoft/teamsfx-core/build/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureFlagName = exports.PluginDisplayName = exports.ResourcePlugins = exports.OutlookClientId = exports.OfficeClientId = exports.TeamsClientId = exports.Bicep = exports.HelpLinks = exports.ConstantString = void 0;
class ConstantString {
}
exports.ConstantString = ConstantString;
ConstantString.UTF8Encoding = "utf-8";
ConstantString.DeploymentResourceType = "Microsoft.Resources/deployments";
ConstantString.DeploymentNotFound = "DeploymentNotFound";
ConstantString.rootFolder = "TeamsApps";
class HelpLinks {
}
exports.HelpLinks = HelpLinks;
HelpLinks.WhyNeedProvision = "https://aka.ms/teamsfx/whyneedprovision";
HelpLinks.ArmHelpLink = "https://aka.ms/teamsfx-arm-help";
class Bicep {
}
exports.Bicep = Bicep;
Bicep.ParameterOrchestrationFileName = "param.template.bicep";
Bicep.ModuleOrchestrationFileName = "module.template.bicep";
Bicep.OutputOrchestrationFileName = "output.template.bicep";
Bicep.VariablesOrchestrationFileName = "variables.template.bicep";
Bicep.ParameterFileName = "parameters.json";
Bicep.ProvisionFileName = "provision.template.bicep";
Bicep.ConfigFileName = "config.template.bicep";
class TeamsClientId {
}
exports.TeamsClientId = TeamsClientId;
TeamsClientId.MobileDesktop = "1fec8e78-bce4-4aaf-ab1b-5451cc387264";
TeamsClientId.Web = "5e3ce6c0-2b1f-4285-8d4b-75ee78787346";
class OfficeClientId {
}
exports.OfficeClientId = OfficeClientId;
OfficeClientId.Web1 = "4345a7b9-9a63-4910-a426-35363201d503";
OfficeClientId.Web2 = "4765445b-32c6-49b0-83e6-1d93765276ca";
class OutlookClientId {
}
exports.OutlookClientId = OutlookClientId;
OutlookClientId.Desktop = "d3590ed6-52b3-4102-aeff-aad2292ab01c";
OutlookClientId.Web = "00000002-0000-0ff1-ce00-000000000000";
class ResourcePlugins {
}
exports.ResourcePlugins = ResourcePlugins;
ResourcePlugins.Aad = "fx-resource-aad-app-for-teams";
ResourcePlugins.FrontendHosting = "fx-resource-frontend-hosting";
ResourcePlugins.SimpleAuth = "fx-resource-simple-auth";
ResourcePlugins.Bot = "fx-resource-bot";
ResourcePlugins.LocalDebug = "fx-resource-local-debug";
ResourcePlugins.AzureSQL = "fx-resource-azure-sql";
ResourcePlugins.Function = "fx-resource-function";
ResourcePlugins.Identity = "fx-resource-identity";
ResourcePlugins.Apim = "fx-resource-apim";
ResourcePlugins.SPFx = "fx-resource-spfx";
class PluginDisplayName {
}
exports.PluginDisplayName = PluginDisplayName;
PluginDisplayName.Solution = "Teams Toolkit";
PluginDisplayName.SpfxSolution = "SPFx";
class FeatureFlagName {
}
exports.FeatureFlagName = FeatureFlagName;
FeatureFlagName.BicepEnvCheckerEnable = "TEAMSFX_BICEP_ENV_CHECKER_ENABLE";
FeatureFlagName.APIV3 = "TEAMSFX_APIV3";
// This will default to true and this environment is only for tests. It does not expose to user.
FeatureFlagName.InsiderPreview = "__TEAMSFX_INSIDER_PREVIEW";
FeatureFlagName.rootDirectory = "TEAMSFX_ROOT_DIRECTORY";
FeatureFlagName.VSCallingCLI = "VS_CALLING_CLI";
