// @microsoft/teamsfx-core/build/common/samples.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sampleProvider = void 0;
const tslib_1 = require("tslib");
const samples_config_json_1 = tslib_1.__importDefault(require("./samples-config.json"));
class SampleProvider {
    get SampleCollection() {
        if (!this.sampleCollection) {
            const samples = samples_config_json_1.default.samples.map((sample) => {
                return {
                    id: sample.id,
                    title: sample.title,
                    shortDescription: sample.shortDescription,
                    fullDescription: sample.fullDescription,
                    tags: sample.tags,
                    time: sample.time,
                    configuration: sample.configuration,
                    link: samples_config_json_1.default.defaultPackageLink,
                };
            });
            this.sampleCollection = {
                baseUrl: samples_config_json_1.default.baseUrl,
                samples,
            };
        }
        return this.sampleCollection;
    }
}
exports.sampleProvider = new SampleProvider();
