// @microsoft/teamsfx-core/build/common/deps-checker/constant/helpLink.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ngrokInstallHelpLink = exports.dotnetNotSupportTargetVersionHelpLink = exports.dotnetManualInstallHelpLink = exports.dotnetFailToInstallHelpLink = exports.dotnetExplanationHelpLink = exports.nodeNotSupportedForSPFxHelpLink = exports.nodeNotSupportedForFunctionsHelpLink = exports.nodeNotSupportedForAzureHelpLink = exports.nodeNotFoundHelpLink = exports.defaultHelpLink = void 0;
exports.defaultHelpLink = "https://aka.ms/teamsfx-envchecker-help";
exports.nodeNotFoundHelpLink = `${exports.defaultHelpLink}#nodenotfound`;
exports.nodeNotSupportedForAzureHelpLink = `${exports.defaultHelpLink}#nodenotsupportedazure-hosting`;
exports.nodeNotSupportedForFunctionsHelpLink = `${exports.defaultHelpLink}#nodenotsupportedazure-functions`;
exports.nodeNotSupportedForSPFxHelpLink = `${exports.defaultHelpLink}#nodenotsupportedspfx-hosting`;
exports.dotnetExplanationHelpLink = `${exports.defaultHelpLink}#overall`;
exports.dotnetFailToInstallHelpLink = `${exports.defaultHelpLink}#failtoinstalldotnet`;
exports.dotnetManualInstallHelpLink = `${exports.defaultHelpLink}#dotnetnotfound`;
exports.dotnetNotSupportTargetVersionHelpLink = `${exports.defaultHelpLink}#dotnetnotsupporttargetversion`;
exports.ngrokInstallHelpLink = `${exports.defaultHelpLink}#failtoinstallngrok`;
