// @microsoft/teamsfx-core/build/plugins/resource/localdebug/util/error.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidLocalBotEndpointFormat = exports.LocalBotEndpointNotConfigured = exports.NgrokTunnelNotConnected = exports.MissingStep = exports.MissingComponent = exports.UnsupportedPlatform = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
function UnsupportedPlatform(platform) {
    return teamsfx_api_1.returnSystemError(new Error(`Unsupported platform: ${platform}.`), "localdebug-plugin", "UnsupportedPlatform");
}
exports.UnsupportedPlatform = UnsupportedPlatform;
function MissingComponent(component) {
    return teamsfx_api_1.returnUserError(new Error(`Component ${component} is required for local debug.`), "localdebug-plugin", "MissingComponent");
}
exports.MissingComponent = MissingComponent;
function MissingStep(operation, requiredStep) {
    return teamsfx_api_1.returnUserError(new Error(`Step "${requiredStep}" is required before ${operation}. Run the required step first.`), "localdebug-plugin", "MissingStep");
}
exports.MissingStep = MissingStep;
function NgrokTunnelNotConnected() {
    return teamsfx_api_1.returnUserError(new Error("Ngrok tunnel is not connected. Check your network settings and try again."), "localdebug-plugin", "NgrokTunnelNotConnected", "https://aka.ms/teamsfx-localdebug");
}
exports.NgrokTunnelNotConnected = NgrokTunnelNotConnected;
function LocalBotEndpointNotConfigured() {
    return teamsfx_api_1.returnUserError(new Error('Local bot endpoint is not configured. Set "fx-resource-local-debug.localBotEndpoint" in ".fx/default.user.data" and try again.'), "localdebug-plugin", "LocalBotEndpointNotConfigured");
}
exports.LocalBotEndpointNotConfigured = LocalBotEndpointNotConfigured;
function InvalidLocalBotEndpointFormat(localBotEndpoint) {
    return teamsfx_api_1.returnUserError(new Error(`Local bot endpoint format is invalid: ${localBotEndpoint}.`), "localdebug-plugin", "InvalidLocalBotEndpointFormat");
}
exports.InvalidLocalBotEndpointFormat = InvalidLocalBotEndpointFormat;
