// @microsoft/teamsfx-core/build/plugins/resource/appstudio/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATIC_TABS_MAX_ITEMS = exports.TEAMS_APP_SHORT_NAME_MAX_LENGTH = exports.DEFAULT_DEVELOPER_PRIVACY_URL = exports.DEFAULT_DEVELOPER_TERM_OF_USE_URL = exports.DEFAULT_DEVELOPER_WEBSITE_URL = exports.BOTS_TPL_LOCAL_DEBUG = exports.COMPOSE_EXTENSIONS_TPL_LOCAL_DEBUG = exports.CONFIGURABLE_TABS_TPL_LOCAL_DEBUG = exports.STATIC_TABS_TPL_LOCAL_DEBUG = exports.BOTS_TPL_FOR_MULTI_ENV = exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV = exports.CONFIGURABLE_TABS_TPL_FOR_MULTI_ENV = exports.STATIC_TABS_TPL_FOR_MULTI_ENV = exports.TEAMS_APP_MANIFEST_TEMPLATE_LOCAL_DEBUG = exports.TEAMS_APP_MANIFEST_TEMPLATE_FOR_MULTI_ENV = exports.SOLUTION_USERINFO = exports.SOLUTION = exports.APP_PACKAGE_FOLDER_FOR_MULTI_ENV = exports.MANIFEST_RESOURCES = exports.DEFAULT_OUTLINE_PNG_FILENAME = exports.DEFAULT_COLOR_PNG_FILENAME = exports.OUTLINE_TEMPLATE = exports.COLOR_TEMPLATE = exports.LOCAL_BOT_ID = exports.BOT_ID = exports.FRONTEND_DOMAIN = exports.FRONTEND_ENDPOINT = exports.MANIFEST_LOCAL = exports.MANIFEST_TEMPLATE = exports.REMOTE_MANIFEST = exports.ErrorMessages = exports.Constants = void 0;
class Constants {
}
exports.Constants = Constants;
Constants.MANIFEST_FILE = "manifest.json";
Constants.PLUGIN_NAME = "AppStudioPlugin";
Constants.PUBLISH_PATH_QUESTION = "manifest-folder";
Constants.BUILD_OR_PUBLISH_QUESTION = "build-or-publish";
Constants.READ_MORE = "Read more";
Constants.LEARN_MORE = "Learn more";
Constants.ADMIN_PORTAL = "Admin portal";
Constants.PUBLISH_GUIDE = "https://aka.ms/teamsfx-publish";
Constants.TEAMS_ADMIN_PORTAL = "https://aka.ms/teamsfx-mtac";
Constants.TEAMS_MANAGE_APP_DOC = "https://aka.ms/teamsfx-mtac-doc";
Constants.TEAMS_APP_ID = "teamsAppId";
Constants.TEAMS_APP_UPDATED_AT = "teamsAppUpdatedAt";
Constants.PERMISSIONS = {
    name: "Teams App",
    noPermission: "No permission",
    admin: "Administrator",
    operative: "Operative",
    type: "M365",
};
Constants.CORRELATION_ID = "X-Correlation-ID";
class ErrorMessages {
}
exports.ErrorMessages = ErrorMessages;
ErrorMessages.GetConfigError = (configName, plugin) => `Failed to get configuration value "${configName}" for ${plugin}.`;
ErrorMessages.ParseUserInfoError = "Failed to parse userInfo.";
ErrorMessages.GrantPermissionFailed = "Response is empty or user is not added.";
ErrorMessages.TeamsAppNotFound = (teamsAppId) => `Cannot find Teams App with id: ${teamsAppId}. Maybe your current M365 account doesn't not have permission, or the Teams App has already been deleted.`;
/**
 * Config keys that are useful for generating remote teams app manifest
 */
exports.REMOTE_MANIFEST = "manifest.source.json";
exports.MANIFEST_TEMPLATE = "manifest.remote.template.json";
exports.MANIFEST_LOCAL = "manifest.local.template.json";
exports.FRONTEND_ENDPOINT = "endpoint";
exports.FRONTEND_DOMAIN = "domain";
exports.BOT_ID = "botId";
exports.LOCAL_BOT_ID = "localBotId";
exports.COLOR_TEMPLATE = "plugins/resource/appstudio/defaultIcon.png";
exports.OUTLINE_TEMPLATE = "plugins/resource/appstudio/defaultOutline.png";
exports.DEFAULT_COLOR_PNG_FILENAME = "color.png";
exports.DEFAULT_OUTLINE_PNG_FILENAME = "outline.png";
exports.MANIFEST_RESOURCES = "resources";
exports.APP_PACKAGE_FOLDER_FOR_MULTI_ENV = "templates/appPackage";
/**
 * Config Keys that are useful for remote collaboration
 */
exports.SOLUTION = "solution";
exports.SOLUTION_USERINFO = "userinfo";
exports.TEAMS_APP_MANIFEST_TEMPLATE_FOR_MULTI_ENV = `{
  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.11/MicrosoftTeams.schema.json",
  "manifestVersion": "1.11",
  "version": "1.0.0",
  "id": "{{state.fx-resource-appstudio.teamsAppId}}",
  "packageName": "com.microsoft.teams.extension",
  "developer": {
      "name": "Teams App, Inc.",
      "websiteUrl": "{{{state.fx-resource-frontend-hosting.endpoint}}}",
      "privacyUrl": "{{{state.fx-resource-frontend-hosting.endpoint}}}/index.html#/privacy",
      "termsOfUseUrl": "{{{state.fx-resource-frontend-hosting.endpoint}}}/index.html#/termsofuse"
  },
  "icons": {
      "color": "resources/color.png",
      "outline": "resources/outline.png"
  },
  "name": {
      "short": "{{config.manifest.appName.short}}",
      "full": "{{config.manifest.appName.full}}"
  },
  "description": {
      "short": "Short description of {{config.manifest.appName.short}}",
      "full": "Full description of {{config.manifest.appName.short}}"
  },
  "accentColor": "#FFFFFF",
  "bots": [],
  "composeExtensions": [],
  "configurableTabs": [],
  "staticTabs": [],
  "permissions": [
      "identity",
      "messageTeamMembers"
  ],
  "validDomains": [],
  "webApplicationInfo": {
      "id": "{{state.fx-resource-aad-app-for-teams.clientId}}",
      "resource": "{{{state.fx-resource-aad-app-for-teams.applicationIdUris}}}"
  }
}`;
exports.TEAMS_APP_MANIFEST_TEMPLATE_LOCAL_DEBUG = `{
  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.11/MicrosoftTeams.schema.json",
  "manifestVersion": "1.11",
  "version": "1.0.0",
  "id": "{{localSettings.teamsApp.teamsAppId}}",
  "packageName": "com.microsoft.teams.extension",
  "developer": {
      "name": "Teams App, Inc.",
      "websiteUrl": "{{{localSettings.frontend.tabEndpoint}}}",
      "privacyUrl": "{{{localSettings.frontend.tabEndpoint}}}/index.html#/privacy",
      "termsOfUseUrl": "{{{localSettings.frontend.tabEndpoint}}}/index.html#/termsofuse"
  },
  "icons": {
      "color": "resources/color.png",
      "outline": "resources/outline.png"
  },
  "name": {
      "short": "{appName}",
      "full": "{appName}"
  },
  "description": {
      "short": "Short description of {appName}",
      "full": "Full description of {appName}"
  },
  "accentColor": "#FFFFFF",
  "bots": [],
  "composeExtensions": [],
  "configurableTabs": [],
  "staticTabs": [],
  "permissions": [
      "identity",
      "messageTeamMembers"
  ],
  "validDomains": [],
  "webApplicationInfo": {
      "id": "{{localSettings.auth.clientId}}",
      "resource": "{{{localSettings.auth.applicationIdUris}}}"
  }
}`;
exports.STATIC_TABS_TPL_FOR_MULTI_ENV = [
    {
        entityId: "index",
        name: "Personal Tab",
        contentUrl: "{{{state.fx-resource-frontend-hosting.endpoint}}}/index.html#/tab",
        websiteUrl: "{{{state.fx-resource-frontend-hosting.endpoint}}}/index.html#/tab",
        scopes: ["personal"],
    },
];
exports.CONFIGURABLE_TABS_TPL_FOR_MULTI_ENV = [
    {
        configurationUrl: "{{{state.fx-resource-frontend-hosting.endpoint}}}/index.html#/config",
        canUpdateConfiguration: true,
        scopes: ["team", "groupchat"],
    },
];
exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        commands: [
            {
                id: "createCard",
                context: ["compose"],
                description: "Command to run action to create a Card from Compose Box",
                title: "Create Card",
                type: "action",
                parameters: [
                    {
                        name: "title",
                        title: "Card title",
                        description: "Title for the card",
                        inputType: "text",
                    },
                    {
                        name: "subTitle",
                        title: "Subtitle",
                        description: "Subtitle for the card",
                        inputType: "text",
                    },
                    {
                        name: "text",
                        title: "Text",
                        description: "Text for the card",
                        inputType: "textarea",
                    },
                ],
            },
            {
                id: "shareMessage",
                context: ["message"],
                description: "Test command to run action on message context (message sharing)",
                title: "Share Message",
                type: "action",
                parameters: [
                    {
                        name: "includeImage",
                        title: "Include Image",
                        description: "Include image in Hero Card",
                        inputType: "toggle",
                    },
                ],
            },
            {
                id: "searchQuery",
                context: ["compose", "commandBox"],
                description: "Test command to run query",
                title: "Search",
                type: "query",
                parameters: [
                    {
                        name: "searchQuery",
                        title: "Search Query",
                        description: "Your search query",
                        inputType: "text",
                    },
                ],
            },
        ],
        messageHandlers: [
            {
                type: "link",
                value: {
                    domains: ["*.botframework.com"],
                },
            },
        ],
    },
];
exports.BOTS_TPL_FOR_MULTI_ENV = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
        commandLists: [
            {
                scopes: ["personal", "team", "groupchat"],
                commands: [
                    {
                        title: "welcome",
                        description: "Resend welcome card of this Bot",
                    },
                    {
                        title: "learn",
                        description: "Learn about Adaptive Card and Bot Command",
                    },
                ],
            },
        ],
    },
];
exports.STATIC_TABS_TPL_LOCAL_DEBUG = [
    {
        entityId: "index",
        name: "Personal Tab",
        contentUrl: "{{{localSettings.frontend.tabEndpoint}}}/index.html#/tab",
        websiteUrl: "{{{localSettings.frontend.tabEndpoint}}}/index.html#/tab",
        scopes: ["personal"],
    },
];
exports.CONFIGURABLE_TABS_TPL_LOCAL_DEBUG = [
    {
        configurationUrl: "{{{localSettings.frontend.tabEndpoint}}}/index.html#/config",
        canUpdateConfiguration: true,
        scopes: ["team", "groupchat"],
    },
];
exports.COMPOSE_EXTENSIONS_TPL_LOCAL_DEBUG = [
    {
        botId: "{{localSettings.bot.botId}}",
        commands: [
            {
                id: "createCard",
                context: ["compose"],
                description: "Command to run action to create a Card from Compose Box",
                title: "Create Card",
                type: "action",
                parameters: [
                    {
                        name: "title",
                        title: "Card title",
                        description: "Title for the card",
                        inputType: "text",
                    },
                    {
                        name: "subTitle",
                        title: "Subtitle",
                        description: "Subtitle for the card",
                        inputType: "text",
                    },
                    {
                        name: "text",
                        title: "Text",
                        description: "Text for the card",
                        inputType: "textarea",
                    },
                ],
            },
            {
                id: "shareMessage",
                context: ["message"],
                description: "Test command to run action on message context (message sharing)",
                title: "Share Message",
                type: "action",
                parameters: [
                    {
                        name: "includeImage",
                        title: "Include Image",
                        description: "Include image in Hero Card",
                        inputType: "toggle",
                    },
                ],
            },
            {
                id: "searchQuery",
                context: ["compose", "commandBox"],
                description: "Test command to run query",
                title: "Search",
                type: "query",
                parameters: [
                    {
                        name: "searchQuery",
                        title: "Search Query",
                        description: "Your search query",
                        inputType: "text",
                    },
                ],
            },
        ],
        messageHandlers: [
            {
                type: "link",
                value: {
                    domains: ["*.botframework.com"],
                },
            },
        ],
    },
];
exports.BOTS_TPL_LOCAL_DEBUG = [
    {
        botId: "{{localSettings.bot.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
        commandLists: [
            {
                scopes: ["personal", "team", "groupchat"],
                commands: [
                    {
                        title: "welcome",
                        description: "Resend welcome card of this Bot",
                    },
                    {
                        title: "learn",
                        description: "Learn about Adaptive Card and Bot Command",
                    },
                ],
            },
        ],
    },
];
// Default values for the developer fields in manifest.
exports.DEFAULT_DEVELOPER_WEBSITE_URL = "https://www.example.com";
exports.DEFAULT_DEVELOPER_TERM_OF_USE_URL = "https://www.example.com/termofuse";
exports.DEFAULT_DEVELOPER_PRIVACY_URL = "https://www.example.com/privacy";
exports.TEAMS_APP_SHORT_NAME_MAX_LENGTH = 30;
exports.STATIC_TABS_MAX_ITEMS = 16;
