// @microsoft/teamsfx-core/build/plugins/resource/appstudio/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppStudioPlugin = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const plugin_1 = require("./plugin");
const constants_1 = require("./constants");
const errors_1 = require("./errors");
const results_1 = require("./results");
const questions_1 = require("./questions");
const telemetry_1 = require("./utils/telemetry");
const typedi_1 = require("typedi");
const ResourcePluginContainer_1 = require("../../solution/fx-solution/ResourcePluginContainer");
const constants_2 = require("../bot/constants");
require("./v2");
const manifestTemplate_1 = require("./manifestTemplate");
let AppStudioPlugin = class AppStudioPlugin {
    constructor() {
        this.name = "fx-resource-appstudio";
        this.displayName = "App Studio";
        this.appStudioPluginImpl = new plugin_1.AppStudioPluginImpl();
    }
    activate(solutionSettings) {
        return true;
    }
    async getQuestions(stage, ctx) {
        var _a;
        const appStudioQuestions = new teamsfx_api_1.QTreeNode({
            type: "group",
        });
        if (stage === teamsfx_api_1.Stage.publish) {
            if (((_a = ctx.answers) === null || _a === void 0 ? void 0 : _a.platform) === teamsfx_api_1.Platform.VSCode) {
                const buildOrPublish = new teamsfx_api_1.QTreeNode({
                    name: constants_1.Constants.BUILD_OR_PUBLISH_QUESTION,
                    type: "singleSelect",
                    staticOptions: [questions_1.manuallySubmitOption, questions_1.autoPublishOption],
                    title: "Teams: Publish to Teams",
                    default: questions_1.autoPublishOption.id,
                });
                appStudioQuestions.addChild(buildOrPublish);
            }
        }
        return teamsfx_api_1.ok(appStudioQuestions);
    }
    /**
     * Create or update teams app
     * For cli: "teamsfx init" only
     * @returns {string} - Remote teams app id
     */
    async getAppDefinitionAndUpdate(ctx, isLocalDebug, manifest) {
        return await this.appStudioPluginImpl.getAppDefinitionAndUpdate(ctx, isLocalDebug, manifest);
    }
    /**
     * Create teams app
     * @returns {string} - Remote teams app id
     */
    async provision(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.provision);
        const remoteTeamsAppId = await this.appStudioPluginImpl.provision(ctx);
        if (remoteTeamsAppId.isErr()) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.provision, remoteTeamsAppId.error, this.appStudioPluginImpl.commonProperties);
            return remoteTeamsAppId;
        }
        else {
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.provision, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.ok(remoteTeamsAppId.value);
        }
    }
    /**
     * Update teams app
     * @returns {string} - Remote teams app id
     */
    async postProvision(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.postProvision);
        const remoteTeamsAppId = await this.appStudioPluginImpl.postProvision(ctx);
        if (remoteTeamsAppId.isErr()) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.postProvision, remoteTeamsAppId.error, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.err(remoteTeamsAppId.error);
        }
        const result = await this.buildTeamsPackage(ctx, false);
        if (result.isErr()) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.postProvision, result.error, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.err(result.error);
        }
        telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.postProvision, this.appStudioPluginImpl.commonProperties);
        return remoteTeamsAppId;
    }
    /**
     * Validate manifest string against schema
     * @param {string} manifestString - the string of manifest.json file
     * @returns {string[]} an array of errors
     */
    async validateManifest(ctx) {
        var _a, _b;
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.validateManifest);
        const validationpluginResult = await this.appStudioPluginImpl.validateManifest(ctx, false);
        if (validationpluginResult.isErr()) {
            return teamsfx_api_1.err(validationpluginResult.error);
        }
        const validationResult = validationpluginResult.value;
        if (validationResult.length > 0) {
            const errMessage = errors_1.AppStudioError.ValidationFailedError.message(validationResult);
            (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.error("Manifest Validation failed!");
            const properties = this.appStudioPluginImpl.commonProperties;
            properties[telemetry_1.TelemetryPropertyKey.validationResult] = validationResult.join("\n");
            const validationFailed = results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.ValidationFailedError.name, errMessage);
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.validateManifest, validationFailed, properties);
            return teamsfx_api_1.err(validationFailed);
        }
        const validationSuccess = "Manifest Validation succeed!";
        (_b = ctx.ui) === null || _b === void 0 ? void 0 : _b.showMessage("info", validationSuccess, false);
        telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.validateManifest, this.appStudioPluginImpl.commonProperties);
        return validationpluginResult;
    }
    async scaffold(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.scaffold);
        try {
            const scaffoldResult = await this.appStudioPluginImpl.scaffold(ctx);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.scaffold);
            return teamsfx_api_1.ok(scaffoldResult);
        }
        catch (error) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.scaffold, error);
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.ScaffoldFailedError.name, errors_1.AppStudioError.ScaffoldFailedError.message(error)));
        }
    }
    /**
     * Build Teams Package
     * @param {string} appDirectory - The directory contains manifest.source.json and two images
     * @returns {string} - Path of built appPackage.zip
     */
    async buildTeamsPackage(ctx, isLocalDebug) {
        var _a;
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.buildTeamsPackage);
        try {
            const appPackagePath = await this.appStudioPluginImpl.buildTeamsAppPackage(ctx, isLocalDebug);
            const builtSuccess = [
                { content: "(√)Done: ", color: teamsfx_api_1.Colors.BRIGHT_GREEN },
                { content: "Teams Package ", color: teamsfx_api_1.Colors.BRIGHT_WHITE },
                { content: appPackagePath, color: teamsfx_api_1.Colors.BRIGHT_MAGENTA },
                { content: " built successfully!", color: teamsfx_api_1.Colors.BRIGHT_WHITE },
            ];
            (_a = ctx.ui) === null || _a === void 0 ? void 0 : _a.showMessage("info", builtSuccess, false);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.buildTeamsPackage, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.ok(appPackagePath);
        }
        catch (error) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.buildTeamsPackage, error, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.TeamsPackageBuildError.name, errors_1.AppStudioError.TeamsPackageBuildError.message(error), error.helpLink));
        }
    }
    /**
     * Migrate V1 project
     */
    async migrateV1Project(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.migrateV1Project);
        try {
            const v1ProjectProperties = await this.appStudioPluginImpl.migrateV1Project(ctx);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.migrateV1Project, {
                enableAuth: v1ProjectProperties.enableAuth ? "true" : "false",
            });
            return teamsfx_api_1.ok(teamsfx_api_1.Void);
        }
        catch (error) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.migrateV1Project, error);
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.MigrateV1ProjectFailedError.name, errors_1.AppStudioError.MigrateV1ProjectFailedError.message(error)));
        }
    }
    /**
     * Update manifest file
     */
    async updateManifest(ctx, isLocalDebug) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.updateManifest);
        const res = await this.appStudioPluginImpl.updateManifest(ctx, isLocalDebug);
        if (res.isErr()) {
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.updateManifest, res.error, this.appStudioPluginImpl.commonProperties);
            if (res.error.name === errors_1.AppStudioError.UpdateManifestCancelError.name) {
                return teamsfx_api_1.ok(teamsfx_api_1.Void);
            }
            else {
                return teamsfx_api_1.err(res.error);
            }
        }
        else {
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.updateManifest, this.appStudioPluginImpl.commonProperties);
            return teamsfx_api_1.ok(teamsfx_api_1.Void);
        }
    }
    /**
     * Publish the app to Teams App Catalog
     * @param {PluginContext} ctx
     * @returns {string[]} - Teams App ID in Teams app catalog
     */
    async publish(ctx) {
        var _a, _b, _c, _d;
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.publish);
        if (((_a = ctx.answers) === null || _a === void 0 ? void 0 : _a.platform) === teamsfx_api_1.Platform.VSCode) {
            const answer = ctx.answers[constants_1.Constants.BUILD_OR_PUBLISH_QUESTION];
            if (answer === questions_1.manuallySubmitOption.id) {
                //const appDirectory = `${ctx.root}/.${ConfigFolderName}`;
                try {
                    const appPackagePath = await this.appStudioPluginImpl.buildTeamsAppPackage(ctx, false);
                    const msg = `Successfully created ${ctx.projectSettings.appName} app package file at ${appPackagePath}. Send this to your administrator for approval.`;
                    (_b = ctx.ui) === null || _b === void 0 ? void 0 : _b.showMessage("info", msg, false, "OK", constants_1.Constants.READ_MORE).then((value) => {
                        var _a;
                        if (value.isOk() && value.value === constants_1.Constants.READ_MORE) {
                            (_a = ctx.ui) === null || _a === void 0 ? void 0 : _a.openUrl(constants_1.Constants.PUBLISH_GUIDE);
                        }
                    });
                    telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.publish);
                    return teamsfx_api_1.ok(appPackagePath);
                }
                catch (error) {
                    telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.publish, error);
                    return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.TeamsPackageBuildError.name, errors_1.AppStudioError.TeamsPackageBuildError.message(error), error.helpLink));
                }
            }
        }
        try {
            const result = await this.appStudioPluginImpl.publish(ctx);
            (_c = ctx.logProvider) === null || _c === void 0 ? void 0 : _c.info(`Publish success!`);
            (_d = ctx.ui) === null || _d === void 0 ? void 0 : _d.showMessage("info", `Success: ${result.name} successfully published to the [admin portal](${constants_1.Constants.TEAMS_ADMIN_PORTAL}). Once approved, your app will be available for your organization.`, false, constants_1.Constants.LEARN_MORE, constants_1.Constants.ADMIN_PORTAL).then((value) => {
                var _a, _b;
                if (value.isOk() && value.value === constants_1.Constants.LEARN_MORE) {
                    (_a = ctx.ui) === null || _a === void 0 ? void 0 : _a.openUrl(constants_1.Constants.TEAMS_MANAGE_APP_DOC);
                }
                else if (value.isOk() && value.value === constants_1.Constants.ADMIN_PORTAL) {
                    (_b = ctx.ui) === null || _b === void 0 ? void 0 : _b.openUrl(constants_1.Constants.TEAMS_ADMIN_PORTAL);
                }
            });
            const properties = this.appStudioPluginImpl.commonProperties;
            properties[telemetry_1.TelemetryPropertyKey.updateExistingApp] = String(result.update);
            properties[telemetry_1.TelemetryPropertyKey.publishedAppId] = String(result.id);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.publish, properties);
            return teamsfx_api_1.ok(result.id);
        }
        catch (error) {
            if (error instanceof teamsfx_api_1.SystemError || error instanceof teamsfx_api_1.UserError) {
                if (error.name === errors_1.AppStudioError.TeamsAppPublishCancelError.name) {
                    telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.publish);
                    return teamsfx_api_1.ok(undefined);
                }
                const innerError = error.innerError ? `innerError: ${error.innerError}` : "";
                error.message = `${error.message} ${innerError}`;
                telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.publish, error, this.appStudioPluginImpl.commonProperties);
                return teamsfx_api_1.err(error);
            }
            else {
                const publishFailed = new teamsfx_api_1.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, error.message, constants_1.Constants.PLUGIN_NAME, undefined, undefined, error);
                telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.publish, publishFailed, this.appStudioPluginImpl.commonProperties);
                return teamsfx_api_1.err(publishFailed);
            }
        }
    }
    async postLocalDebug(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.localDebug);
        const localTeamsAppId = await this.appStudioPluginImpl.postLocalDebug(ctx);
        if (localTeamsAppId.isOk()) {
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.localDebug);
            await this.appStudioPluginImpl.buildTeamsAppPackage(ctx, true);
            return localTeamsAppId;
        }
        else {
            const error = localTeamsAppId.error;
            if (error instanceof teamsfx_api_1.SystemError || error instanceof teamsfx_api_1.UserError) {
                telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.localDebug, error);
                return teamsfx_api_1.err(error);
            }
            else {
                const updateFailedError = results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.LocalAppIdUpdateFailedError.name, errors_1.AppStudioError.LocalAppIdUpdateFailedError.message(error));
                telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.localDebug, updateFailedError);
                return teamsfx_api_1.err(updateFailedError);
            }
        }
    }
    async checkPermission(ctx, userInfo) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.checkPermission);
        try {
            const checkPermissionResult = await this.appStudioPluginImpl.checkPermission(ctx, userInfo);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.checkPermission);
            return teamsfx_api_1.ok(checkPermissionResult);
        }
        catch (error) {
            const fxError = error.name && error.name >= 400 && error.name < 500
                ? results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.CheckPermissionFailedError.name, errors_1.AppStudioError.CheckPermissionFailedError.message(error))
                : results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CheckPermissionFailedError.name, errors_1.AppStudioError.CheckPermissionFailedError.message(error));
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.checkPermission, fxError);
            return teamsfx_api_1.err(fxError);
        }
    }
    async grantPermission(ctx, userInfo) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.grantPermission);
        try {
            const grantPermissionResult = await this.appStudioPluginImpl.grantPermission(ctx, userInfo);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.grantPermission);
            return teamsfx_api_1.ok(grantPermissionResult);
        }
        catch (error) {
            const fxError = error.name && error.name >= 400 && error.name < 500
                ? results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.GrantPermissionFailedError.name, errors_1.AppStudioError.GrantPermissionFailedError.message(error.message))
                : results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.GrantPermissionFailedError.name, errors_1.AppStudioError.GrantPermissionFailedError.message(error.message));
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.grantPermission, fxError);
            return teamsfx_api_1.err(fxError);
        }
    }
    async listCollaborator(ctx) {
        telemetry_1.TelemetryUtils.init(ctx);
        telemetry_1.TelemetryUtils.sendStartEvent(telemetry_1.TelemetryEventName.listCollaborator);
        try {
            const listCollaborator = await this.appStudioPluginImpl.listCollaborator(ctx);
            telemetry_1.TelemetryUtils.sendSuccessEvent(telemetry_1.TelemetryEventName.listCollaborator);
            return teamsfx_api_1.ok(listCollaborator);
        }
        catch (error) {
            const fxError = error.name && error.name >= 400 && error.name < 500
                ? results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.ListCollaboratorFailedError.name, errors_1.AppStudioError.ListCollaboratorFailedError.message(error))
                : results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.ListCollaboratorFailedError.name, errors_1.AppStudioError.ListCollaboratorFailedError.message(error));
            telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.listCollaborator, fxError);
            return teamsfx_api_1.err(fxError);
        }
    }
    async loadManifest(ctx) {
        const localManifest = await manifestTemplate_1.loadManifest(ctx.root, true);
        if (localManifest.isErr()) {
            return teamsfx_api_1.err(localManifest.error);
        }
        const remoteManifest = await manifestTemplate_1.loadManifest(ctx.root, false);
        if (remoteManifest.isErr()) {
            return teamsfx_api_1.err(remoteManifest.error);
        }
        return teamsfx_api_1.ok({ local: localManifest.value, remote: remoteManifest.value });
    }
    async saveManifest(ctx, manifest) {
        let res = await manifestTemplate_1.saveManifest(ctx.root, manifest.local, true);
        if (res.isErr()) {
            return teamsfx_api_1.err(res.error);
        }
        res = await manifestTemplate_1.saveManifest(ctx.root, manifest.remote, false);
        if (res.isErr()) {
            return teamsfx_api_1.err(res.error);
        }
        return teamsfx_api_1.ok(undefined);
    }
    async capabilityExceedLimit(ctx, capability) {
        return await manifestTemplate_1.capabilityExceedLimit(ctx.root, capability);
    }
    async init(ctx) {
        return await manifestTemplate_1.init(ctx.root);
    }
    async addCapabilities(ctx, capabilities) {
        return await manifestTemplate_1.addCapabilities(ctx.root, capabilities);
    }
    async executeUserTask(func, ctx) {
        if (func.method === "validateManifest") {
            return await this.validateManifest(ctx);
        }
        else if (func.method === "buildPackage") {
            if (func.params && func.params.type) {
                const isLocalDebug = func.params.type === "localDebug";
                return await this.buildTeamsPackage(ctx, isLocalDebug);
            }
            return await this.buildTeamsPackage(ctx, false);
        }
        else if (func.method === "getAppDefinitionAndUpdate") {
            if (func.params && func.params.type && func.params.manifest) {
                const isLocalDebug = func.params.type === "localDebug";
                return await this.getAppDefinitionAndUpdate(ctx, isLocalDebug, func.params.manifest);
            }
            return teamsfx_api_1.err(new teamsfx_api_1.SystemError("InvalidParam", `Invalid param:${JSON.stringify(func)}`, constants_1.Constants.PLUGIN_NAME, undefined, constants_2.Links.ISSUE_LINK));
        }
        else if (func.method === "migrateV1Project") {
            return await this.migrateV1Project(ctx);
        }
        else if (func.method === "getManifestTemplatePath") {
            const isLocalDebug = func.params.type === "localDebug";
            const filePath = await manifestTemplate_1.getManifestTemplatePath(ctx.root, isLocalDebug);
            return teamsfx_api_1.ok(filePath);
        }
        else if (func.method === "updateManifest") {
            return await this.updateManifest(ctx, func.params && func.params.envName === "local");
        }
        return teamsfx_api_1.err(new teamsfx_api_1.SystemError(constants_1.Constants.PLUGIN_NAME, "FunctionRouterError", `Failed to route function call:${JSON.stringify(func)}`, constants_2.Links.ISSUE_LINK));
    }
};
AppStudioPlugin = tslib_1.__decorate([
    typedi_1.Service(ResourcePluginContainer_1.ResourcePlugins.AppStudioPlugin)
], AppStudioPlugin);
exports.AppStudioPlugin = AppStudioPlugin;
exports.default = new AppStudioPlugin();
