// @microsoft/teamsfx-core/build/plugins/solution/fx-solution/utils/progressHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressHelper = exports.DeployArmTemplatesSteps = void 0;
const common_1 = require("../../../../common");
var DeployArmTemplatesSteps;
(function (DeployArmTemplatesSteps) {
    DeployArmTemplatesSteps["ExecuteDeployment"] = "Deploying solution ARM templates to Azure. This could take several minutes.";
})(DeployArmTemplatesSteps = exports.DeployArmTemplatesSteps || (exports.DeployArmTemplatesSteps = {}));
class ProgressHelper {
    static async startDeployArmTemplatesProgressHandler(ui) {
        var _a, _b;
        await ((_a = this.deployArmTemplatesProgress) === null || _a === void 0 ? void 0 : _a.end(true));
        this.deployArmTemplatesProgress = ui === null || ui === void 0 ? void 0 : ui.createProgressBar(common_1.getStrings().solution.DeployArmTemplates.Progress.Title, Object.entries(DeployArmTemplatesSteps).length);
        await ((_b = this.deployArmTemplatesProgress) === null || _b === void 0 ? void 0 : _b.start(common_1.getStrings().solution.DeployArmTemplates.Progress.Start));
        return this.deployArmTemplatesProgress;
    }
    static async endDeployArmTemplatesProgress(success) {
        var _a;
        await ((_a = this.deployArmTemplatesProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.deployArmTemplatesProgress = undefined;
    }
}
exports.ProgressHelper = ProgressHelper;
