// @microsoft/teamsfx-core/build/plugins/solution/fx-solution/v3/solution.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsFxAzureSolution = void 0;
const tslib_1 = require("tslib");
const typedi_1 = require("typedi");
const addModule_1 = require("./addModule");
const addResource_1 = require("./addResource");
const constants_1 = require("./constants");
const deploy_1 = require("./deploy");
const init_1 = require("./init");
const provision_1 = require("./provision");
const provisionLocal_1 = require("./provisionLocal");
const publish_1 = require("./publish");
const scaffold_1 = require("./scaffold");
const userTask_1 = require("./userTask");
let TeamsFxAzureSolution = class TeamsFxAzureSolution {
    constructor() {
        this.name = constants_1.TeamsFxAzureSolutionNameV3;
        this.getQuestionsForInit = init_1.getQuestionsForInit;
        this.init = init_1.init;
        this.getQuestionsForScaffold = scaffold_1.getQuestionsForScaffold;
        this.scaffold = scaffold_1.scaffold;
        this.getQuestionsForAddResource = addResource_1.getQuestionsForAddResource;
        this.addResource = addResource_1.addResource;
        this.getQuestionsForAddModule = addModule_1.getQuestionsForAddModule;
        this.addModule = addModule_1.addModule;
        this.getQuestionsForProvision = provision_1.getQuestionsForProvision;
        this.provisionResources = provision_1.provisionResources.bind(this);
        this.getQuestionsForLocalProvision = provisionLocal_1.getQuestionsForLocalProvision;
        this.provisionLocalResources = provisionLocal_1.provisionLocalResources;
        this.getQuestionsForDeploy = deploy_1.getQuestionsForDeploy;
        this.deploy = deploy_1.deploy;
        this.getQuestionsForPublish = publish_1.getQuestionsForPublish;
        this.publishApplication = publish_1.publishApplication;
        this.getQuestionsForUserTask = userTask_1.getQuestionsForUserTask;
        this.executeUserTask = userTask_1.executeUserTask;
    }
};
TeamsFxAzureSolution = tslib_1.__decorate([
    typedi_1.Service(constants_1.TeamsFxAzureSolutionNameV3)
], TeamsFxAzureSolution);
exports.TeamsFxAzureSolution = TeamsFxAzureSolution;
