// @microsoft/teamsfx-core/build/common/deps-checker/depsError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackendExtensionsInstallError = exports.LinuxNotSupportedError = exports.NodeNotSupportedError = exports.NodeNotFoundError = exports.DepsCheckerError = void 0;
class DepsCheckerError extends Error {
    constructor(message, helpLink) {
        super(message);
        this.helpLink = helpLink;
        Object.setPrototypeOf(this, DepsCheckerError.prototype);
    }
}
exports.DepsCheckerError = DepsCheckerError;
class NodeNotFoundError extends DepsCheckerError {
    constructor(message, helpLink) {
        super(message, helpLink);
        Object.setPrototypeOf(this, NodeNotFoundError.prototype);
    }
}
exports.NodeNotFoundError = NodeNotFoundError;
class NodeNotSupportedError extends DepsCheckerError {
    constructor(message, helpLink) {
        super(message, helpLink);
        Object.setPrototypeOf(this, NodeNotSupportedError.prototype);
    }
}
exports.NodeNotSupportedError = NodeNotSupportedError;
class LinuxNotSupportedError extends DepsCheckerError {
    constructor(helpLink) {
        super("Linux is not supported.", helpLink);
        Object.setPrototypeOf(this, LinuxNotSupportedError.prototype);
    }
}
exports.LinuxNotSupportedError = LinuxNotSupportedError;
class BackendExtensionsInstallError extends DepsCheckerError {
    constructor(message, helpLink) {
        super(message, helpLink);
        Object.setPrototypeOf(this, BackendExtensionsInstallError.prototype);
    }
}
exports.BackendExtensionsInstallError = BackendExtensionsInstallError;
